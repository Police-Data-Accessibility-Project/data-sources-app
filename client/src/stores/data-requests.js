import { defineStore } from 'pinia';

export const useDataRequestsStore = defineStore('data-requests', {
	state: () => ({
		/** Searches performed during session. */
		cache: {},
		/** Previous route visited - useful for determining whether we are incrementing or decrementing pages in data request by id */
		previousDataRequestRoute: null,
	}),
	getters: {
		getDataRequestFromCache: (state) => (key) => {
			if (!(key in state.cache)) {
				return null;
			}
			return state.cache[key];
		},
	},

	actions: {
		clearCache() {
			this.$patch((state) => {
				state.cache = {};
			});
		},
		setDataRequestToCache(key, data, timestamp = new Date().getTime()) {
			this.$patch((state) => {
				state.cache[key] = {
					data,
					timestamp,
				};
			});
		},
		setPreviousDataRequestRoute(route) {
			this.$patch({
				previousDataSourceRoute: route,
			});
		},
	},
});
