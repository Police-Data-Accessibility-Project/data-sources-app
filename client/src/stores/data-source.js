import { defineStore } from 'pinia';

export const useDataSourceStore = defineStore('data-source', {
	state: () => ({
		/** Previous route visited - useful for determining whether we are incrementing or decrementing pages in data source by id */
		previousDataSourceRoute: null,
		/** Cache by ID -- useful for avoiding unnecessary API calls */
		cache: {},
	}),
	persist: {
		storage: sessionStorage,
	},
	getters: {
		getDataSourceFromCache: (state) => (key) => {
			if (!(key in state.cache)) {
				return null;
			}
			return state.cache[key];
		},
	},
	actions: {
		clearCache() {
			this.$patch((state) => {
				state.cache = {};
			});
		},

		setPreviousDataSourceRoute(route) {
			this.$patch({
				previousDataSourceRoute: route,
			});
		},

		setDataSourceToCache(key, data) {
			this.$patch((state) => {
				state.cache[key] = {
					data,
					timestamp: new Date().getTime(),
				};
			});
		},
	},
});
