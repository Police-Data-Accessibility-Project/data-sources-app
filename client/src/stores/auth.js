import axios from 'axios';
import { defineStore } from 'pinia';
import parseJwt from '../util/parseJwt';
import router from '../router';
import { useUserStore } from './user';

const HEADERS = {
	headers: { 'Content-Type': 'application/json' },
};
const LOGIN_URL = `${import.meta.env.VITE_VUE_API_BASE_URL}/login`;
const REFRESH_SESSION_URL = `${import.meta.env.VITE_VUE_API_BASE_URL}/refresh-session`;

export const useAuthStore = defineStore('auth', {
	state: () => ({
		userId: null,
		accessToken: {
			value: null,
			expires: Date.now(),
		},
		returnUrl: null,
	}),
	persist: true,
	actions: {
		async login(email, password) {
			const user = useUserStore();

			try {
				const response = await axios.post(
					LOGIN_URL,
					{ email, password },
					HEADERS,
				);

				// Update user store with email
				user.$patch({ email });

				this.parseTokenAndSetData(response);
				if (this.returnUrl) router.push(this.returnUrl);
			} catch (error) {
				throw new Error(error.response?.data?.message);
			}
		},

		logout(isAuthRoute) {
			const user = useUserStore();

			this.$patch({
				userId: null,
				accessToken: { value: null, expires: Date.now() },
				returnUrl: null,
			});

			user.$patch({
				email: '',
			});
			if (isAuthRoute) router.push('/login');
		},

		async refreshAccessToken() {
			if (!this.$state.userId) return;
			try {
				const response = await axios.post(
					REFRESH_SESSION_URL,
					{ session_token: this.$state.accessToken.value },
					HEADERS,
				);
				return this.parseTokenAndSetData(response);
			} catch (error) {
				throw new Error(error.response?.data?.message);
			}
		},

		parseTokenAndSetData(response) {
			const token = response.data.data;
			const tokenParsed = parseJwt(token);

			this.$patch({
				userId: tokenParsed.sub,
				accessToken: {
					value: token,
					expires: new Date(tokenParsed.exp * 1000).getTime(),
				},
			});
		},
	},
});
