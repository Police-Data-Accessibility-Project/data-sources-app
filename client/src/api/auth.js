import axios from 'axios';
import { ENDPOINTS } from './constants';
import { useAuthStore } from '@/stores/auth';
import { useUserStore } from '@/stores/user';

const AUTH_BASE = `${import.meta.env.VITE_VUE_API_BASE_URL}/auth`;
const HEADERS = {
	'Content-Type': 'application/json',
};

export async function signUpWithEmail(email, password) {
	const user = useUserStore();

	await axios.post(
		`${AUTH_BASE}/${ENDPOINTS.AUTH.SIGNUP}`,
		{ email, password },
		{ headers: HEADERS },
	);
	// Update store with email
	user.$patch({ email });
}

export async function signInWithEmail(email, password) {
	const auth = useAuthStore();

	const response = await axios.post(
		`${AUTH_BASE}/${ENDPOINTS.AUTH.LOGIN}`,
		{ email, password },
		{
			headers: {
				...HEADERS,
				// TODO: API should require auth
				// authorization: `Basic ${import.meta.env.VITE_ADMIN_API_KEY}`,
			},
		},
	);

	auth.parseTokensAndSetData(response);
}

export async function beginOAuthLogin(redirectPath = '/sign-in') {
	const redirectTo = encodeURI(
		`${AUTH_BASE}/${ENDPOINTS.OAUTH.GITHUB}?redirect_url=${import.meta.env.VITE_VUE_APP_BASE_URL}${redirectPath}`,
	);

	window.location.href = redirectTo;
}

export async function signInWithGithub(gh_access_token) {
	const auth = useAuthStore();

	const response = await axios.post(
		`${AUTH_BASE}/${ENDPOINTS.OAUTH.LOGIN_WITH_GITHUB}`,
		{ gh_access_token },
		{
			headers: {
				...HEADERS,
			},
		},
	);

	auth.parseTokensAndSetData(response);
	return true;
}

export async function linkAccountWithGithub(gh_access_token) {
	const { email: user_email } = useUserStore();

	return await axios.post(
		`${AUTH_BASE}/${ENDPOINTS.OAUTH.LINK_TO_GITHUB}`,
		{ gh_access_token, user_email },
		{
			headers: {
				...HEADERS,
				// authorization: `Bearer ${this.$state.tokens.accessToken.value}`,
			},
		},
	);
}

export async function signOut() {
	const auth = useAuthStore();
	const user = useUserStore();

	auth.$reset();
	user.$reset();
}

export async function refreshTokens() {
	const auth = useAuthStore();
	const user = useUserStore();

	if (!user.id) return;
	try {
		const response = await axios.post(
			`${AUTH_BASE}/${ENDPOINTS.AUTH.REFRESH_SESSION}`,
			{ refresh_token: auth.$state.tokens.refreshToken.value },
			{
				headers: {
					...HEADERS,
					authorization: `Bearer ${auth.$state.tokens.accessToken.value}`,
				},
			},
		);
		return auth.parseTokensAndSetData(response);
	} catch (error) {
		console.error(error);
	}
}

export async function validateEmail(token) {
	const auth = useAuthStore();

	const response = await axios.post(
		`${AUTH_BASE}/${ENDPOINTS.AUTH.VALIDATE_EMAIL}`,
		null,
		{
			headers: {
				...HEADERS,
				Authorization: `Bearer ${token}`,
			},
		},
	);

	auth.parseTokensAndSetData(response);
}

export async function resendValidationEmail() {
	const { email } = useUserStore();

	return await axios.post(
		`${AUTH_BASE}/${ENDPOINTS.AUTH.RESEND_VALIDATION_EMAIL}`,
		{ email },
		{
			headers: {
				...HEADERS,
			},
		},
	);
}

export async function requestPasswordReset(email) {
	return await axios.post(
		`${AUTH_BASE}/${ENDPOINTS.AUTH.REQUEST_RESET_PASSWORD}`,
		{ email },
		{ headers: HEADERS },
	);
}

export async function resetPassword(password, token) {
	return await axios.post(
		`${AUTH_BASE}/${ENDPOINTS.AUTH.RESET_PASSWORD}`,
		{ password },
		{ headers: { ...HEADERS, Authorization: 'Bearer ' + token } },
	);
}

/**
 * @deprecated validation now done by parsing JWT directly
 */
export async function validateResetPasswordToken(token) {
	return await axios.post(
		`${AUTH_BASE}/${ENDPOINTS.AUTH.RESET_TOKEN_VALIDATION}`,
		{
			headers: {
				...HEADERS,
				Authorization: `Bearer ${token}`,
			},
		},
	);
}
