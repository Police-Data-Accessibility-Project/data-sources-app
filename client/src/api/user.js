import axios from 'axios';
import { ENDPOINTS } from './constants';
import { useAuthStore } from '@/stores/auth';
import { useUserStore } from '@/stores/user';

const USER_BASE = `${import.meta.env.VITE_VUE_API_BASE_URL}/user`;
const HEADERS = {
	'Content-Type': 'application/json',
};

export async function changePassword(oldPassword, newPassword) {
	const auth = useAuthStore();
	const user = useUserStore();

	await axios.post(
		`${USER_BASE}/${user.id}/${ENDPOINTS.USER.ID.UPDATE_PASSWORD}`,
		{ old_password: oldPassword, new_password: newPassword },
		{
			headers: {
				...HEADERS,
				Authorization: `Bearer ${auth.$state.tokens.accessToken.value}`,
			},
		},
	);

	return await auth.signInWithEmail(user.email, newPassword);
}
