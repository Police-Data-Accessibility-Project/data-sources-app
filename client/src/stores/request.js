import { useAuthStore } from '@/stores/auth';
import axios from 'axios';
import { defineStore } from 'pinia';

// Constants
const HEADERS_BASE = {
	authorization: `Basic ${import.meta.env.VITE_ADMIN_API_KEY}`,
	'Content-Type': 'application/json',
};
const REQUESTS_URL = `${import.meta.env.VITE_VUE_API_BASE_URL}/data-requests`;

export const useRequestStore = defineStore('request', {
	state: () => ({}),
	// persist: {
	// 	storage: sessionStorage,
	// },
	actions: {
		async createRequest(data) {
			const { tokens } = useAuthStore();
			const response = await axios.post(REQUESTS_URL, data, {
				headers: {
					...HEADERS_BASE,
					authorization: `Bearer ${tokens.accessToken.value}`,
				},
			});

			return response.data;
		},
	},
});
