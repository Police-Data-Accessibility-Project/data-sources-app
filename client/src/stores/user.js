import axios from 'axios';
import { defineStore } from 'pinia';
import { useAuthStore } from './auth';

const HEADERS = {
	headers: { 'Content-Type': 'application/json' },
};
const SIGNUP_URL = `${import.meta.env.VITE_VUE_API_BASE_URL}/user`;
const CHANGE_PASSWORD_URL = `${import.meta.env.VITE_VUE_API_BASE_URL}/user`;
const REQUEST_PASSWORD_RESET_URL = `${import.meta.env.VITE_VUE_API_BASE_URL}/request-reset-password`;
const PASSWORD_RESET_URL = `${import.meta.env.VITE_VUE_API_BASE_URL}/reset-password`;

export const useUserStore = defineStore('user', {
	state: () => ({
		email: '',
	}),
	persist: true,
	actions: {
		async signup(email, password) {
			const auth = useAuthStore();

			try {
				await axios.post(SIGNUP_URL, { email, password }, HEADERS);
				// Update store with email
				this.$patch({ email });
				// Log users in after signup and return that response
				return await auth.login(email, password);
			} catch (error) {
				throw new Error(error.response?.data?.message);
			}
		},

		async changePassword(email, password) {
			const auth = useAuthStore();
			try {
				await axios.put(
					CHANGE_PASSWORD_URL,
					{ email, password },
					{
						headers: {
							...HEADERS.headers,
							Authorization: `Bearer ${auth.accessToken.value}`,
						},
					},
				);
				return await auth.login(email, password);
			} catch (error) {
				throw new Error(error.response?.data?.message);
			}
		},

		async requestPasswordReset(email) {
			try {
				await axios.post(REQUEST_PASSWORD_RESET_URL, { email }, HEADERS);
			} catch (error) {
				throw new Error(error.response?.data?.message);
			}
		},

		async resetPassword(password, token) {
			try {
				await axios.post(`${PASSWORD_RESET_URL}`, { password, token }, HEADERS);
			} catch (error) {
				throw new Error(error.response?.data?.message);
			}
		},
	},
});
