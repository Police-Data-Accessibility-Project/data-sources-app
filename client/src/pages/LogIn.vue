<template>
	<!-- User is already logged in -->
	<main v-if="auth.userId" class="pdap-flex-container">
		<h1 data-test="success-heading">
			{{ success ? 'Success' : "You're already logged in" }}
		</h1>
		<p data-test="success-subheading">
			{{ success ? success : 'Enjoy the data sources app!' }}
		</p>

		<Button
			class="mt-6"
			data-test="logout-button"
			@click="
				() => {
					auth.logout();
					success = '';
				}
			"
			>Log out</Button
		>
	</main>

	<!-- Otherwise, the form (form handles error UI on its own) -->
	<main v-else class="pdap-flex-container mx-auto max-w-2xl">
		<h1>Sign In</h1>
		<Form
			id="login"
			class="flex flex-col"
			data-test="login-form"
			name="login"
			:error="error"
			:schema="FORM_SCHEMAS[type]"
			@change="handleChangeOnError"
			@submit="onSubmit"
		>
			<ul v-if="type === FORM_TYPES.signup" class="text-med mb-8">
				Passwords must be at least 8 characters and include:
				<li>1 uppercase letter</li>
				<li>1 lowercase letter</li>
				<li>1 number</li>
				<li>1 special character</li>
			</ul>

			<Button class="max-w-full" type="submit" data-test="submit-button">
				{{ getSubmitButtonCopy() }}
			</Button>
		</Form>
		<div
			class="flex flex-col items-start sm:flex-row sm:items-center sm:gap-4 w-full"
		>
			<Button
				class="flex-1 max-w-full"
				intent="secondary"
				data-test="toggle-button"
				@click="toggleType"
			>
				{{ type === FORM_TYPES.login ? 'Create Account' : 'Log In' }}
			</Button>
			<RouterLink
				class="pdap-button-secondary flex-1 max-w-full"
				data-test="reset-link"
				to="/reset-password"
				>Reset Password</RouterLink
			>
		</div>
	</main>
</template>

<script setup>
// Imports
import { Button, Form } from 'pdap-design-system';
import { ref, watchEffect } from 'vue';
import { useAuthStore } from '../stores/auth';
import { useUserStore } from '../stores/user';

// Constants
const LOGIN_SCHEMA = [
	{
		'data-test': 'email',
		id: 'email',
		name: 'email',
		label: 'Email',
		type: 'text',
		placeholder: 'Your email address',
		value: '',
		validators: {
			email: {
				message: 'Please provide your email address',
				value: true,
			},
		},
	},
	{
		'data-test': 'password',
		id: 'password',
		name: 'password',
		label: 'Password',
		type: 'password',
		placeholder: 'Your password',
		value: '',
		validators: {
			password: {
				message: 'Please provide your password',
				value: true,
			},
		},
	},
];

// Signing up we want to validate that PW matches, so we add that field
const SIGNUP_SCHEMA = LOGIN_SCHEMA.concat({
	'data-test': 'confirm-password',
	id: 'confirmPassword',
	name: 'confirmPassword',
	label: 'Confirm Password',
	type: 'password',
	placeholder: 'Confirm your password',
	value: '',
	validators: {
		password: {
			message: 'Please confirm your password',
			value: true,
		},
	},
});

const SUCCESS_COPY = {
	login: "You're now logged in!",
	signup: 'Your account is now active!',
};

/**
 * Enum-like object for TS stans ðŸ˜Š
 */
const FORM_TYPES = {
	login: 'login',
	signup: 'signup',
};

const FORM_SCHEMAS = {
	[FORM_TYPES.login]: LOGIN_SCHEMA,
	[FORM_TYPES.signup]: SIGNUP_SCHEMA,
};

// Store
const auth = useAuthStore();
const user = useUserStore();

// Reactive vars
const error = ref(undefined);
const loading = ref(false);
const success = ref(undefined);
const type = ref(FORM_TYPES.login);

// Effects
// Clear error on success
watchEffect(() => {
	if (success.value) error.value = undefined;
});

// Functions
// Handlers
/**
 * When signing up: handles clearing pw-match form errors on change if they exist
 */
function handleChangeOnError(formValues) {
	if (
		type.value === FORM_TYPES.signup &&
		error.value &&
		formValues.password !== formValues.confirmPassword
	) {
		handlePasswordValidation(formValues);
	}
}

/**
 * When signing up: validates that passwords match
 * @returns {boolean} `false` if passwords do not match, `true` if they do
 */
function handlePasswordValidation(formValues) {
	if (formValues.password !== formValues.confirmPassword) {
		if (!error.value) {
			error.value = 'Passwords do not match, please try again.';
		}
		return false;
	} else {
		error.value = undefined;
		return true;
	}
}

/**
 * Logs user in or signs user up
 */
async function onSubmit(formValues) {
	if (type.value === FORM_TYPES.signup) {
		if (!handlePasswordValidation(formValues)) return;
	}

	try {
		loading.value = true;
		const { email, password } = formValues;

		type.value === FORM_TYPES.signup
			? await user.signup(email, password)
			: await auth.login(email, password);

		success.value = SUCCESS_COPY[type.value];
	} catch (err) {
		error.value = err.message;
	} finally {
		loading.value = false;
	}
}

// Utils
/**
 * Toggles between login and signup actions
 */
function toggleType() {
	switch (type.value) {
		case FORM_TYPES.login:
			type.value = FORM_TYPES.signup;
			break;
		case FORM_TYPES.signup:
		default:
			type.value = FORM_TYPES.login;
			break;
	}
}

function getSubmitButtonCopy() {
	switch (true) {
		case loading.value:
			return 'Loading...';
		case type.value === FORM_TYPES.signup:
			return 'Create account';
		case type.value === FORM_TYPES.login:
		default:
			return 'Login';
	}
}
</script>
