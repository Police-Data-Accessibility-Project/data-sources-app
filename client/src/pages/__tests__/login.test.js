import { RouterLinkStub, flushPromises, mount } from '@vue/test-utils';
import { describe, expect, vi, beforeEach, it } from 'vitest';
import { nextTick } from 'vue';
import LogIn from '../LogIn.vue';
import { createTestingPinia } from '@pinia/testing';
import { useAuthStore } from '../../stores/auth';
import { useUserStore } from '../../stores/user';

const push = vi.fn();
const $routerMock = {
	push,
};

let wrapper;

describe('Login page', () => {
	beforeEach(() => {
		wrapper = mount(LogIn, {
			global: {
				mocks: {
					router: $routerMock,
				},
				plugins: [createTestingPinia()],
				stubs: {
					RouterLink: RouterLinkStub,
				},
			},
		});
	});

	describe('Login/Logout', () => {
		let auth;

		beforeEach(() => {
			auth = useAuthStore();
		});

		it('Calls the login method with valid form data', async () => {
			const email = wrapper.find('[data-test="email"] input');
			const password = wrapper.find('[data-test="password"] input');
			const form = wrapper.find('[data-test="login-form"]');

			await email.setValue('hello@hello.com');
			await password.setValue('Password1!');

			form.trigger('submit');
			await flushPromises();

			expect(auth.login).toHaveBeenCalledWith('hello@hello.com', 'Password1!');
			expect(wrapper.html()).toMatchSnapshot();
		});

		describe('Success and already logged in states', async () => {
			beforeEach(() => {
				auth.userId = 42;
			});

			it('Displays success copy', async () => {
				wrapper.vm.success = "You're now logged in!";
				await nextTick();

				const heading = await wrapper.find('[data-test="success-heading"]');
				const subheading = await wrapper.find(
					'[data-test="success-subheading"]',
				);

				expect(heading.text()).toBe('Success');
				expect(subheading.text()).toBe("You're now logged in!");
			});

			it('Logs user out', async () => {
				const logout = await wrapper.find('[data-test="logout-button"]');

				expect(logout.exists()).toBe(true);

				logout.trigger('click');
				await flushPromises();

				expect(auth.logout).toHaveBeenCalledOnce();
			});
		});

		it('Handles API error', async () => {
			const mockError = new Error('foo');
			vi.mocked(auth.login).mockRejectedValueOnce(mockError);

			const email = wrapper.find('[data-test="email"] input');
			const password = wrapper.find('[data-test="password"] input');
			const form = wrapper.find('[data-test="login-form"]');

			await email.setValue('hello@hello.com');
			await password.setValue('Password1!');

			form.trigger('submit');
			await flushPromises();

			const error = form.find('.pdap-form-error-message');
			expect(error.exists()).toBe(true);
			expect(error.text()).toBe('foo');
		});
	});

	describe('Signup', () => {
		let toggle;
		let user;

		beforeEach(() => {
			toggle = wrapper.find('[data-test="toggle-button"]');
			user = useUserStore();
		});

		it('Calls the signup method with valid data', async () => {
			expect(toggle.exists()).toBe(true);
			toggle.trigger('click');

			await nextTick();

			const email = wrapper.find('[data-test="email"] input');
			const password = wrapper.find('[data-test="password"] input');
			const confirmPassword = wrapper.find(
				'[data-test="confirm-password"] input',
			);
			const form = wrapper.find('[data-test="login-form"]');

			await email.setValue('hello@hello.com');
			await password.setValue('Password1!');
			await confirmPassword.setValue('Password1!');

			form.trigger('submit');
			await flushPromises();
			await nextTick();

			expect(user.signup).toHaveBeenCalledOnce();
			expect(wrapper.html()).toMatchSnapshot();
		});

		it('Renders error message with mismatched passwords when trying to sign up and re-validates form', async () => {
			expect(toggle.exists()).toBe(true);
			toggle.trigger('click');

			await nextTick();

			const email = wrapper.find('[data-test="email"] input');
			const password = wrapper.find('[data-test="password"] input');
			const confirmPassword = wrapper.find(
				'[data-test="confirm-password"] input',
			);
			const form = wrapper.find('[data-test="login-form"]');

			await email.setValue('hello@hello.com');
			await password.setValue('Password1!');
			await confirmPassword.setValue('Password');

			form.trigger('submit');
			await flushPromises();
			await nextTick();

			const error = form.find('.pdap-form-error-message');

			expect(error.exists()).toBe(true);
			expect(error.text()).toBe('Passwords do not match, please try again.');
			expect(wrapper.html()).toMatchSnapshot();

			await nextTick();

			confirmPassword.setValue('Pasdasdfasdf');
			await nextTick();
			expect(error.exists()).toBe(true);
		});
	});

	describe('Miscellaneous states', () => {
		it('Toggles form type', async () => {
			const toggle = wrapper.find('[data-test="toggle-button"]');
			const submit = wrapper.find('[data-test="submit-button"]');

			toggle.trigger('click');
			await nextTick();

			expect(submit.text()).toBe('Create account');

			toggle.trigger('click');
			await nextTick();

			expect(submit.text()).toBe('Login');
		});

		it('Renders button loading copy', async () => {
			wrapper.vm.loading = true;
			wrapper.vm.$forceUpdate();

			expect(wrapper.vm.getSubmitButtonCopy()).toBe('Loading...');
		});
	});
});
