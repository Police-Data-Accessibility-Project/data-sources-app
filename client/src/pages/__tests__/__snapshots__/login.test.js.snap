// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login page > Login/Logout > Calls the login method with valid form data 1`] = `
<!-- User is already logged in -->
<!-- Otherwise, the form (form handles error UI on its own) -->
<main class="pdap-flex-container mx-auto max-w-2xl">
  <h1>Sign In</h1>
  <form class="pdap-form flex flex-col" id="login" name="login">
    <!---->
    <div class="pdap-grid-container pdap-input">
      <input id="email" name="email" placeholder="Your email address" type="text">
      <!---->
      <!---->
      <label class="pdap-input-label" for="email">Email</label>
    </div>
    <div class="pdap-grid-container pdap-input">
      <!---->
      <input id="password" name="password" placeholder="Your password" type="password">
      <!---->
      <label class="pdap-input-label" for="password">Password</label>
    </div>
    <!--v-if-->
    <button class="pdap-button pdap-button-primary max-w-full" type="submit">Login</button>
  </form>
  <div class="flex flex-col items-start sm:flex-row sm:items-center sm:gap-4 w-full">
    <button class="pdap-button pdap-button-secondary flex-1 max-w-full">Create Account</button>
    <a class="pdap-button-secondary flex-1 max-w-full">Reset Password</a>
  </div>
</main>
`;

exports[`Login page > Signup > Calls the signup method with valid data 1`] = `
<!-- User is already logged in -->
<!-- Otherwise, the form (form handles error UI on its own) -->
<main class="pdap-flex-container mx-auto max-w-2xl">
  <h1>Sign In</h1>
  <form class="pdap-form flex flex-col" id="login" name="login">
    <!---->
    <div class="pdap-grid-container pdap-input">
      <input id="email" name="email" placeholder="Your email address" type="text">
      <!---->
      <!---->
      <label class="pdap-input-label" for="email">Email</label>
    </div>
    <div class="pdap-grid-container pdap-input">
      <!---->
      <input id="password" name="password" placeholder="Your password" type="password">
      <!---->
      <label class="pdap-input-label" for="password">Password</label>
    </div>
    <div class="pdap-grid-container pdap-input">
      <!---->
      <input id="confirmPassword" name="confirmPassword" placeholder="Confirm your password" type="password">
      <!---->
      <label class="pdap-input-label" for="confirmPassword">Confirm Password</label>
    </div>
    <ul class="text-med mb-8"> Passwords must be at least 8 characters and include: <li>1 uppercase letter</li>
      <li>1 lowercase letter</li>
      <li>1 number</li>
      <li>1 special character</li>
    </ul>
    <button class="pdap-button pdap-button-primary max-w-full" type="submit">Create account</button>
  </form>
  <div class="flex flex-col items-start sm:flex-row sm:items-center sm:gap-4 w-full">
    <button class="pdap-button pdap-button-secondary flex-1 max-w-full">Log In</button>
    <a class="pdap-button-secondary flex-1 max-w-full">Reset Password</a>
  </div>
</main>
`;

exports[`Login page > Signup > Renders error message with mismatched passwords when trying to sign up and re-validates form 1`] = `
<!-- User is already logged in -->
<!-- Otherwise, the form (form handles error UI on its own) -->
<main class="pdap-flex-container mx-auto max-w-2xl">
  <h1>Sign In</h1>
  <form class="pdap-form flex flex-col" id="login" name="login">
    <div class="pdap-form-error-message">Passwords do not match, please try again.</div>
    <div class="pdap-grid-container pdap-input">
      <input id="email" name="email" placeholder="Your email address" type="text">
      <!---->
      <!---->
      <label class="pdap-input-label" for="email">Email</label>
    </div>
    <div class="pdap-grid-container pdap-input">
      <!---->
      <input id="password" name="password" placeholder="Your password" type="password">
      <!---->
      <label class="pdap-input-label" for="password">Password</label>
    </div>
    <div class="pdap-grid-container pdap-input">
      <!---->
      <input id="confirmPassword" name="confirmPassword" placeholder="Confirm your password" type="password">
      <!---->
      <label class="pdap-input-label" for="confirmPassword">Confirm Password</label>
    </div>
    <ul class="text-med mb-8"> Passwords must be at least 8 characters and include: <li>1 uppercase letter</li>
      <li>1 lowercase letter</li>
      <li>1 number</li>
      <li>1 special character</li>
    </ul>
    <button class="pdap-button pdap-button-primary max-w-full" type="submit">Create account</button>
  </form>
  <div class="flex flex-col items-start sm:flex-row sm:items-center sm:gap-4 w-full">
    <button class="pdap-button pdap-button-secondary flex-1 max-w-full">Log In</button>
    <a class="pdap-button-secondary flex-1 max-w-full">Reset Password</a>
  </div>
</main>
`;
